include/rocksdb/env.h

```diff
@@ -1339,253 +1345,297 @@ extern Status ReadFileToString(Env* env, const std::string& fname,
 // functionality of another Env.
 class EnvWrapper : public Env {
  public:
+  // The Target struct allows an Env to be stored as a raw (Env*) or
+  // std::shared_ptr<Env>.  By using this struct, the wrapping/calling
+  // class does not need to worry about the ownership/lifetime of the
+  // wrapped target env.  If the guard is set, then the Env will point
+  // to the guard.get().
+  struct Target {
+    Env* env;                    // The raw Env
+    std::shared_ptr<Env> guard;  // The guarded Env
```

Comment:
Should we also create a unique_ptr<Env> member to hold the target env? Using shared_ptr is kind of strange.

Reply:
The value has to be stored as a shared_ptr, because that is what CreateFromString can return.  CreateFromString has to return a shared_ptr, as it is possible the same Env is shared between multiple DBs.

And it is not all that unusual to create a shared from a unique (but you cannot do it the other way around).

Reply:
Ok. Maybe indicate in the comment that the initialization from both `unique_ptr` and `shared_ptr` will be maintained by this "guard"